### PR TITLE
AEGIS-X: include the parallelism degree in the context

### PR DESCRIPTION
**Pull request by jedisct1**
This is a breaking change, and it is not strictly necessary, but since it is not too late, it may be desirable.

Currently, if a (key,nonce) is reused with different variants of the same security level, the key streams initialized with the same context will be the same.

For example, with AEGIS-128L and AEGIS-128X2, the 128L key stream will be the same as the first half of the 128X2 key stream.

Tags will be different, but reusing a (key,nonce) with different variants would still reveal the difference between the messages.

We can document that (key,nonce) pairs must not be reused with different algorithms.

But a better alternative would be to include the degree in the context.

We actually use (degree - 1), so that when the degree is 1, the context remains all-zero, and it doesn't break the equivalence between AEGIS-128L and AEGIS-128X1.

### CODE REVIEW

**Comment by jedisct1** [comment]
If you agree with that change, the paper will be updated as well.

**Comment by jedisct1** [comment]
Hi @samuel-lucas6 , whenever it's convenient for you, could you take a look at this? It would be great to tag a new version with the parallel variants.

**Comment by samuel-lucas6** [comment]
Sorry Frank, I was on holiday. I'll take a look tomorrow morning. It sounds sensible, and it's a good idea to also update the paper.

**Comment by jedisct1** [comment]
No worries Samuel! Vacations are more important than anything 😊

File: draft-irtf-cfrg-aegis-aead.md [hunk 1]
```diff
@@ -979,7 +979,7 @@ The state of a parallel mode is represented as a vector of AEGIS-128L or AEGIS-2
 - `R`: the absorption and output rate of the mode. With AEGIS-128X, the rate is `2 * 128 * D` bits. With AEGIS-256X, the rate is `128 * D` bits.
 - `V[j,i]`: the `j`-th AES block of the `i`-th state. `i` is in the `[0..D)` range. For AEGIS-128X, `j` is in the `[0..8)` range, while for AEGIS-256, `j` is in the `[0..6)` range.
 - `V'[j,i]`: the `j`-th AES block of the next `i`-th state.
-- `ctx`: the context separator.
+- `ctx[i]`: the `i`-th context separator: a 128-bit mask, made of a byte representing the state index, followed by a byte representing the highest index and 112 all-zero bits.
```
Comment (samuel-lucas6) [inline draft-irtf-cfrg-aegis-aead.md]: It's perhaps worth rephrasing this without the colon due to the colon on the left.

Comment (samuel-lucas6) [reply draft-irtf-cfrg-aegis-aead.md]: Could be `context separator. This is a 128-bit mask...` or similar.

**Review by samuel-lucas6 — CHANGES_REQUESTED** [review]
Looks good. One potential change I can see. The `ZeroPad()` once is an improvement.

File: draft-irtf-cfrg-aegis-aead.md (hunk 1 above)
Comment (jedisct1) [reply draft-irtf-cfrg-aegis-aead.md]: It sounds indeed better without the column. Your suggestion was merged. Thank you Samuel!

**Comment by jedisct1** [comment]
The `ZeroPad` repetition was originally there because the context itself was the byte, not the padding.

But that didn't really match the way it would actually be implemented. Describing it as a mask solved it :)

**Pull request merged**